### PR TITLE
Mesh filename publishing

### PR DESCRIPTION
**Pull request by patmarion**
This addresses the discussion in #1636.

Note, this depends on the ability to check whether a path is relative or absolute. The spruce third party library that drake uses for filename manipulation does not provide this capability.  As a workaround, this pull request implements a check for relative paths that assumes posix paths, it does not work on Windows (so don't merge this yet).

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/2239) &emsp; Multiple assignees:&emsp;@david-german-tri&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/david-german-tri">david-german-tri</a>,&emsp;@patmarion&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/patmarion">patmarion</a>

### CODE REVIEW

**Comment by patmarion** [comment]
[copy and pasted comment from #1721]

In this pull request, mesh filenames are always published with absolute paths. Previously, mesh filenames were published as relative paths or absolute paths depending on whether the urdf command line argument was given as a relative or absolute path.

This pull request also converts prius.urdf to use package:// syntax. This makes lcm logs with DRAKE_VIEWER_LOAD messages portable between computers and verifies that drake-visualizer is handling package:// correctly.

Note, drake-visualizer imports pydrake in order to call pydrake.getDrakePath(). It uses the drake path to recursively search for packages (dirs with package.xml files).

**Comment by patmarion** [comment]
This pull request resolves #1636.  But this feature is blocked until we have a cross-platform implementation of  IsRelativePath() or IsAbsolutePath() in Drake.  I'd like to ask @david-german-tri or @sherm1 if they have any advice?  Thanks!

**Comment by patmarion** [comment]
Assigned to @david-german-tri for comment/advice.

I set priority to low since people seem to be dealing with this issue, it is not blocking at the moment.  The current workaround to is give absolute paths to URDF files on the command line.

**Comment by david-german-tri** [comment]
The interface/approach LGTM.  As you say, I think it's clear we need some kind of cross-platform path library, but I'm the wrong person to ask for suggestions, as I have ~0 Windows experience.  +@sherm1, over to you?

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/BotVisualizer.h, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-KGnUqs_u_eUNAIYBpFJ:-KGnUqs_u_eUNAIYBpFK:-1265811731) ([raw file](https://github.com/robotlocomotion/drake/blob/fee83bea569c16b8282da8048d1869d595fd912c/drake/systems/plants/BotVisualizer.h#L107)):*
This could use test coverage.

---

*[drake/systems/plants/xmlUtil.cpp, line 211 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-KGnSoDBAWnP3s3AsnHl:-KGnSoDBAWnP3s3AsnHm:-171825364) ([raw file](https://github.com/robotlocomotion/drake/blob/fee83bea569c16b8282da8048d1869d595fd912c/drake/systems/plants/xmlUtil.cpp#L211)):*
Windows path support elsewhere (e.g. addRobotFromSDF) is fairly hacky.  Assuming we don't find a decent library: if you wanted to regexp for `[A-Z]:\` or something like that, I wouldn't judge (as long as there was test coverage using an actual file on the filesystem).

---

*[drake/systems/plants/xmlUtil.cpp, line 215 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-KGnRbZobRs2DU1ZDeOS:-KGnRbZp_iAe7ek3C9Pw:1265825785) ([raw file](https://github.com/robotlocomotion/drake/blob/fee83bea569c16b8282da8048d1869d595fd912c/drake/systems/plants/xmlUtil.cpp#L215)):*
To check my understanding: if the filename is something like `../foo/bar/baz.urdf`, root_dir is `./foo/bar/baz`, and we wind up sending something like `/home/user/other/../foo/bar/baz.urdf`, correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-:-KGnR8-438t3PqXUbJPG:-771691889)*

**Comment by sherm1** [comment]
@patmarion: Spruce "normalizes" the path by changing all backslashes to forward slashes. That means the only difference between a Windows path and a Unix one is the possible presence of the drive letter at the beginning. So if you have a method for determining whether a Unix pathname is absolute, you can apply that same method to a Windows pathname after stripping off the drive if it is present. If present, the drive will be `[a-zA-Z]:` in positions 0 and 1 of the string. You should only do that on Windows because those are legitimate file name characters elsewhere. You can reliably determine you are on Windows with `#ifdef _WIN32`. Perversely, that symbol is defined regardless of whether you are running 32- or 64-bit Windows.

**Comment by patmarion** [comment]
I tried adding a commit that uses the Windows API function [PathIsRelative](https://msdn.microsoft.com/en-us/library/windows/desktop/bb773660%28v=vs.85%29.aspx) but it appears that API requires an extra link library.  I don't have access to a Windows dev machine, but this is the commit I added

https://github.com/RobotLocomotion/drake/pull/2239/commits/737530eb81ec9729324c69bdfa9791554d4c58f4

and this is the link error from the jenkins build:

xmlUtil.obj : error LNK2019: unresolved external symbol __imp__PathIsRelativeA@4

I don't know if this Windows API is something we can be using in Drake.  If not, then I guess I'll try the regex test to match Windows drive letters.

**Comment by patmarion** [comment]
By the way, I found that Drake already has a Windows implementation of GetFullPath, but it is provided as a mex function for Matlab.  The Matlab URDF parsers use it.  It's here:

https://github.com/RobotLocomotion/drake/blob/master/drake/thirdParty/GetFullPath/GetFullPath.c

It doesn't look like there is an easy way to call that for my purpose though.

**Comment by sherm1** [comment]
> I don't know if this Windows API is something we can be using in Drake. If not, then I guess I'll try the regex test to match Windows drive letters.

I think adding the Windows Shell library would be overkill for getting this method. You don't have to do this with a regex -- I just wrote it that way for clarity. You can just do something like:
```c++
#include <cctype>
// assume path has been normalized by spruce
std::string driveless = 
  path.size() >= 2 && std::isalpha(path[0]) && path[1]==':' 
  ? path.substr(2, path.size()-2) // trim off drive
  : path;
// see if driveless is absolute or relative
```

**Comment by patmarion** [comment]
Thanks @sherm1 I've used your example code to detect drive letters.

@david-german-tri here is an example of how mesh filenames are resolved in this topic branch:

    # path to urdf given as relative path
    $ ./pod-build/bin/carSimLCM examples/Cars/models/prius/prius.urdf
    
      loading urdf file: examples/Cars/models/prius/prius.urdf
      resolving mesh filename: prius.dae
      normalized_root_dir: examples/Cars/models/prius
      converted relative root dir to absolute: /Users/pat/source/drake/drake/drake/examples/Cars/models/prius
      mesh filename: /Users/pat/source/drake/drake/drake/examples/Cars/models/prius/prius.dae

    # path to urdf given as absolute path
    $ ./pod-build/bin/carSimLCM $PWD/examples/Cars/models/prius/prius.urdf
    
      loading urdf file: /Users/pat/source/drake/drake/drake/examples/Cars/models/prius/prius.urdf
      resolving mesh filename: prius.dae
      root dir: /Users/pat/source/drake/drake/drake/examples/Cars/models/prius
      mesh filename: /Users/pat/source/drake/drake/drake/examples/Cars/models/prius/prius.dae

    # when urdf specifies mesh filename using package syntax
    $ ./pod-build/bin/carSimLCM examples/Cars/models/prius/prius.urdf
    
      loading urdf file: examples/Cars/models/prius/prius.urdf
      resolving mesh filename: package://drake/examples/Cars/models/prius/prius.dae
      mesh filename: /Users/pat/source/drake/drake/drake/examples/Cars/models/prius/prius.dae

**Comment by liangfok** [comment]
While we're at it, should we also support the `model://` format that's supported by Gazebo?

**Comment by patmarion** [comment]
I would recommend we don't add a new feature to this pull request.  This is focused on resolving relative paths of mesh filenames.  Maybe you could open a new issue for model:// url support?

**Comment by liangfok** [comment]
Done: https://github.com/RobotLocomotion/drake/issues/2249

**Comment by patmarion** [comment]
Sigh.  rospack functionality appears to be broken for matlab?  This test fails:

testing /home/ubuntu/workspace/experimental/06107e1d/drake/examples/Cars/models/prius/prius.urdf
Error using rospack (line 51)
couldn't find ROS package drake.
 ROS_ROOT=
 ROS_PACKAGE_PATH=

I will remove the package:// addition I made to prius.urdf.  I added it so that tests could demonstrate package:// support works. It works in C++ but matlab appears to be broken?  Anyway, it's a separate issue, so I will remove it from this pull request.

**Comment by patmarion** [comment]
I removed the package://drake from prius.urdf using a revert commit:

https://github.com/RobotLocomotion/drake/pull/2239/commits/29071b1ea38220875c33cc161df05fcfbfa46ab8

Now all tests should pass.  When the rospack feature is working in Matlab then the compareParsers
test should pass and we can revert the revert commit.  Perhaps the feature works now, but it doesn't automatically search the drake source dir for packages?

**Comment by patmarion** [comment]
The rospack issue:

in c++, populatePackageMap() implemented in xmlUtil recursively searches the drake root, so it picks up the drake/package.xml.  Then it searches the standard environment variables.

in matlab, rospack.m only adds certain package dirs in the examples, it does not recursively search the drake root.  Then it searches the standard environment variables.

@RussTedrake maybe we can extend rospack.m to recursively search getDrakePath() for package.xml, so the implementation matches xmlUtil's populatePackageMap()?

**Comment by patmarion** [comment]
This is ready to merge.  It passed all tests on CI.  I just clicked the Update branch button, so it will be re-tested I guess.  Assigning to @david-german-tri for final review and merge please.

**Comment by david-german-tri** [comment]
The code change looks fine to me, but I renew my request for unit tests.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-:-KGsXJD7sPjdFmLTnjZO:1679809886)*

**Comment by patmarion** [comment]
I believe there are multiple existing tests that exercise these code paths, for example, comparePriusURDFtoSDF loads the prius.urdf with a relative path name.  The code path you highlighted in BotVisualizer.h won't be tested until I can add package:// syntax to prius.urdf, which is blocked by the matlab rospack issue I described above.

**Comment by david-german-tri** [comment]
:lgtm:

---

Reviewed 1 of 4 files at r2, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/systems/plants/BotVisualizer.h, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-KGnUqs_u_eUNAIYBpFJ:-KGsagajz9d8Bx3uQSVU:-1282086481) ([raw file](https://github.com/robotlocomotion/drake/blob/fee83bea569c16b8282da8048d1869d595fd912c/drake/systems/plants/BotVisualizer.h#L107)):*
Resolved, see above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2239#-:-KGsaftooNP3V-I0WUkv:1417088721)*

**Comment by david-german-tri** [comment]
Passed Jenkins prior to roll-up merge, so I'm proceeding with override merge.

**Comment by patmarion** [comment]
thanks for your help guys!

**Comment by jwnimmer-tri** [comment]
I assume this post-merge build was a flake, but we'll keep our eye on the subsequent builds of master just in case: https://drake-jenkins.csail.mit.edu/view/Continuous/job/windows-msvc-32-continuous-debug/

**Comment by jwnimmer-tri** [comment]
The subsequent build 120 of master (after #2246 got merged) also failed in the same way as this PR.
Continuing to investigate.

**Comment by jwnimmer-tri** [comment]
I can't make any sense of it.  Will have to leave to @RobotLocomotion/kitware in the morning.

**Comment by david-german-tri** [comment]
I'm going to try reverting this to see if that fixes the ongoing Windows breakage.

**Comment by patmarion** [comment]
sounds good

File: drake/systems/plants/xmlUtil.cpp [hunk 1]
```diff
@@ -205,7 +205,25 @@ string resolveFilename(const string& filename,
       return string();
     }
   } else {
-    mesh_filename_s = spruce::path(root_dir);
+    std::string normalized_root_dir = spruce::path(root_dir).getStr();
+
+    // if root_dir is a relative path then convert it to absolute
+#ifdef _WIN32
+    bool dirIsRelative = !(normalized_root_dir.size() >= 2
+                           && std::isalpha(normalized_root_dir[0])
```
Comment (patmarion) [inline drake/systems/plants/xmlUtil.cpp]: This is the only line that actually introduces something new, a reference to the std::isalpha function.  The rest of the branch is just std string manipulation.

**Pull request merged**